### ISSUE TITLE
Persistent Data when migrating presents permissions issues

### ISSUE THREAD

**Issue by SavSanta** [bitnami/bitnami-docker-mariadb#263]
**Description**

Im migrating a redmine instance that has a persistent MariaDB volume on an old Linux machine which I tar'd up. I am attempting to migrate that MariaDB persistent volume into a new Linux machine and have untar'd it in the same location. I've edited file permissions on the new Linux to be the same from the old Linux machine. Both are using bitnami mariadb (latest docker image). 

MariaDB will give me permission errors about the copied over persistent volume on the new Linux machine and then Mariadb container usually crashes, wont start/hang with various errors usually with mentions of permissions related. 

I've tried mirorring the 'data' directory permissions of the old Linux machine unto the new Linux machine, but  that doesnt work..........however I noticed that when I chmod the "data" directory to 776 or 777.

A semi-demonstration follows

## Failure Permissions of the data directory

```
root@redmine:/media/nfs/mariadb_data/mariadb/data# ls -l
total 189088
-rwxrw---- 1 redmineuser redmineuser    630784 Feb  3 18:54 aria_log.00000001
-rwxrw---- 1 redmineuser redmineuser        52 Feb  3 18:54 aria_log_control
-rwxrw---- 1 redmineuser redmineuser         9 Feb  3 18:02 ddl_recovery.log
-rwxrw---- 1 redmineuser redmineuser      6165 Feb  3 17:42 ib_buffer_pool
-rwxrw---- 1 redmineuser redmineuser  79691776 Feb  3 18:54 ibdata1
-rwxrw---- 1 redmineuser redmineuser 100663296 Feb  3 18:54 ib_logfile0
-rwxrw---- 1 redmineuser redmineuser  12582912 Feb  3 17:42 ibtmp1
-rwxrw---- 1 redmineuser redmineuser         0 Jul 30  2021 multi-master.info
drwxrw---- 2 redmineuser redmineuser      4096 Feb  3 17:42 mysql
-rwxr--r-- 1 redmineuser redmineuser        15 Feb  3 17:42 mysql_upgrade_info
drwxrw---- 2 redmineuser redmineuser      4096 Feb  3 17:42 performance_schema
drwxrw---- 2 redmineuser redmineuser      4096 Aug  1  2021 redmine_production
drwxrw---- 2 redmineuser redmineuser     12288 Feb  3 17:42 sys
drwxrw---- 2 redmineuser redmineuser      4096 Jul 30  2021 test
```

## Failing/Crashing/Hanging (the latter in this case) MariaDB Snippet

```
2022-02-03 23:02:46 0 [ERROR] InnoDB: The error means mariadbd does not have the access rights to the directory.
2022-02-03 23:02:46 0 [ERROR] InnoDB: Could not find a valid tablespace file for redmine_production/workflows. Please refer to https://mariadb.com/kb/en/innodb-data-dictionary-troubleshooting/ for how to resolve the issue.
2022-02-03 23:02:46 0 [Warning] InnoDB: Ignoring tablespace for redmine_production/workflows because it could not be opened.
2022-02-03 23:02:46 0 [ERROR] InnoDB: Operating system error number 13 in a file operation.
2022-02-03 23:02:46 0 [ERROR] InnoDB: The error means mariadbd does not have the access rights to the directory.
2022-02-03 23:02:46 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2022-02-03 23:02:46 0 [Note] InnoDB: 10.6.5 started; log sequence number 30903125; transaction id 39340
2022-02-03 23:02:46 0 [Note] Plugin 'FEEDBACK' is disabled.
2022-02-03 23:02:46 0 [ERROR] Could not open mysql.plugin table: "Table 'mysql.plugin' doesn't exist". Some plugins may be not loaded
2022-02-03 23:02:46 0 [ERROR] mysqld: Can't create/write to file './ddl_recovery-backup.log' (Errcode: 13 "Permission denied")
2022-02-03 23:02:46 0 [Note] InnoDB: Loading buffer pool(s) from /bitnami/mariadb/data/ib_buffer_pool
2022-02-03 23:02:46 0 [ERROR] Can't open and lock privilege tables: Table 'mysql.servers' doesn't exist
2022-02-03 23:02:46 0 [Note] Server socket created on IP: '127.0.0.1'.
2022-02-03 23:02:46 0 [Note] InnoDB: Buffer pool(s) load completed at 220203 23:02:46
2022-02-03 23:02:46 0 [ERROR] Fatal error: Can't open and lock privilege tables: Table 'mysql.db' doesn't exist
2022-02-03 23:02:46 0 [ERROR] Aborting
2022-02-03 23:02:46 0 [ERROR] InnoDB: Cannot open '/bitnami/mariadb/data/ib_buffer_pool.incomplete' for writing: Permission denied
2022-02-03 23:02:46 0 [ERROR] InnoDB: Operating system error number 13 in a file operation.
2022-02-03 23:02:46 0 [ERROR] InnoDB: The error means mariadbd does not have the access rights to the directory.

```

---

## Successfully Permissions of the data directory with chmod of 776
```
root@redmine:/media/nfs/mariadb_data/mariadb/data# ls -l
total 189088
-rwxrwxrw- 1 redmineuser redmineuser    630784 Feb  3 18:54 aria_log.00000001
-rwxrwxrw- 1 redmineuser redmineuser        52 Feb  3 18:54 aria_log_control
-rwxrwxrw- 1 redmineuser redmineuser         9 Feb  3 18:02 ddl_recovery.log
-rwxrwxrw- 1 redmineuser redmineuser      6165 Feb  3 17:42 ib_buffer_pool
-rwxrwxrw- 1 redmineuser redmineuser  79691776 Feb  3 18:54 ibdata1
-rwxrwxrw- 1 redmineuser redmineuser 100663296 Feb  3 18:54 ib_logfile0
-rwxrwxrw- 1 redmineuser redmineuser  12582912 Feb  3 17:42 ibtmp1
-rwxrwxrw- 1 redmineuser redmineuser         0 Jul 30  2021 multi-master.info
drwxrwxrw- 2 redmineuser redmineuser      4096 Feb  3 17:42 mysql
-rwxrwxrw- 1 redmineuser redmineuser        15 Feb  3 17:42 mysql_upgrade_info
drwxrwxrw- 2 redmineuser redmineuser      4096 Feb  3 17:42 performance_schema
drwxrwxrw- 2 redmineuser redmineuser      4096 Aug  1  2021 redmine_production
drwxrwxrw- 2 redmineuser redmineuser     12288 Feb  3 17:42 sys
drwxrwxrw- 2 redmineuser redmineuser      4096 Jul 30  2021 test
root@redmine:/media/nfs/mariadb_data/mariadb/data# 
```

## Successful Running MariaDB Snippet

```
mariadb 00:01:35.45 INFO  ==> ** Starting MariaDB **
2022-02-04  0:01:35 0 [Note] /opt/bitnami/mariadb/sbin/mysqld (server 10.6.5-MariaDB) starting as process 1 ...
2022-02-04  0:01:35 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2022-02-04  0:01:35 0 [Note] InnoDB: Number of pools: 1
2022-02-04  0:01:35 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
2022-02-04  0:01:35 0 [Note] mysqld: O_TMPFILE is not supported on /opt/bitnami/mariadb/tmp (disabling future attempts)
2022-02-04  0:01:35 0 [Note] InnoDB: Using Linux native AIO
2022-02-04  0:01:35 0 [Note] InnoDB: Initializing buffer pool, total size = 134217728, chunk size = 134217728
2022-02-04  0:01:35 0 [Note] InnoDB: Completed initialization of buffer pool
2022-02-04  0:01:35 0 [Note] InnoDB: 128 rollback segments are active.
2022-02-04  0:01:35 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2022-02-04  0:01:35 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2022-02-04  0:01:35 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
2022-02-04  0:01:35 0 [Note] InnoDB: 10.6.5 started; log sequence number 30903173; transaction id 39340
2022-02-04  0:01:35 0 [Note] Plugin 'FEEDBACK' is disabled.
2022-02-04  0:01:35 0 [Note] InnoDB: Loading buffer pool(s) from /bitnami/mariadb/data/ib_buffer_pool
2022-02-04  0:01:35 0 [Note] Server socket created on IP: '0.0.0.0'.
2022-02-04  0:01:35 0 [Warning] 'proxies_priv' entry '@% root@redmine' ignored in --skip-name-resolve mode.
2022-02-04  0:01:35 0 [Note] /opt/bitnami/mariadb/sbin/mysqld: ready for connections.
Version: '10.6.5-MariaDB'  socket: '/opt/bitnami/mariadb/tmp/mysql.sock'  port: 3306  Source distribution
2022-02-04  0:01:35 0 [Note] InnoDB: Buffer pool(s) load completed at 220204  0:01:35
```

My additional probably off base thoughts:
_I believe it may be an issue to do with the user id. As I should note there is a few differences between the old Linux host machine and the new Linux host machine. The old has two users a power user account with a UID of 1000 and then an account that happens to be equivalent to the "redmineuser" account with a coincidental UID of 1001 on that host machine has a UID of 1001 which is equivalent to the 1001 in the MariaDB container. Maybe related?_

Do let me know if there is anything else I can do to further dig into the matter.

Thanks

**Comment by SavSanta** [comment]
Fuller output from the hanging attempt and the working attempts:

[HangingMaria.txt](https://github.com/bitnami/bitnami-docker-mariadb/files/7999144/HangingMaria.txt)
[WorkingMaria.txt](https://github.com/bitnami/bitnami-docker-mariadb/files/7999145/WorkingMaria.txt)

**Comment by carrodher** [comment]
Please note the bitnami/mariadb container image is non-root, this means it is executed as the user `1001` which belongs to the root group. It should work with any user belonging to the root group.

You can find more info about non-root containers at
- https://docs.bitnami.com/kubernetes/faq/configuration/use-non-root/
- https://docs.bitnami.com/tutorials/bitnami-best-practices-hardening-containers/#root-and-non-root-containers
- https://docs.bitnami.com/tutorials/work-with-non-root-containers/
- https://engineering.bitnami.com/articles/why-non-root-containers-are-important-for-security.html

**Comment by SavSanta** [comment]
It should be noted that Im using podman which uses rootless containers on both Linux installs by default. 

Curiously. I went ahead and created an additional user and group on the Linux host that have UIDs of 1001 and changed the data permissions so that said new user and group own the directory.

Doing this didnt really make any change however. Though it's 230 am here and after struggling with this i need a bit of a nap.

What did work I'll post in a separate comment

**Comment by SavSanta** [comment]
To further clarify for anyone else before closing this. I too am used to docker but previous implementor preferred podman usage due to better security aspect of rootless containers. Thusly, I discovered my issue and I had noticed it before some days ago of this may have something to do with usage of subuid values /etc/subuid as the permissions on the data directory did have what appeared to be a random number that mapped within the range of the UID of the actual [Linux Host] user as the owner. However the manpage for /etc/subuid it is rather laconic and not exploratory, that I presumed I could replace the "random number" with the actual Linux Host username/userid.

Somehow I  after using `ps` on the Linux Host immediately after launching the mariadb container, I caught in the process list a "userid" value that looked high and nonstandard (obviously in hindisght related subid ). Output looked like this

```
   1000   11381  0.0  1.9 1187676 39728 pts/0   Sl+  07:34   0:00 podman run -it --pod redminer --name mariadb -e MARIADB_PASSWORD=test 
    1000   11387  0.0  2.4 1263328 49284 pts/0   Sl+  07:34   0:00 podman run -it --pod redminer --name mariadb -e MARIADB_PASSWORD=test 
    1000   11404  0.0  0.1   6392  3584 ?        Ss   07:34   0:00 /usr/bin/fuse-overlayfs -o ,lowerdir=/home/redmineuser/.local/share/containers/sto
    1000   11406  0.0  0.1  81276  2000 ?        Ssl  07:34   0:00 /usr/bin/conmon --api-version 1 -c 8218ccecc749765e5dfaf60c124624d2e31b38ab28c483c
  101000   11409  0.0  5.2 1062388 105172 pts/0  Ssl+ 07:34   0:01 /opt/bitnami/mariadb/sbin/mysqld --defaults-file=/opt/bitnami/mariadb/conf/my.cnf 
```

I check my /etc/subuid on my Linux Host machine because I remember the number being 100000 and was one of the owners I changed it to a few days ago rather than 1001000. 

```
root@redmine:/media/nfs/mariadb_data/mariadb/data# cat /etc/subuid
redmineuser:100000:65536
```

So I refined my DuckDuckGo/Startpage web searching and ended up on https://blog.christophersmart.com/2021/01/26/user-ids-and-rootless-containers-with-podman/

This page explains it a little bit better about using rootless containers with podman and how the subid works in section "Running rootless containers as root, as you"

Where it to paraphase states "the container root account with uid 0 maps to the 1000 uid of our non-root user on the Linux host.  Then, the uid of 1 in a container for my user would map to 100000 on the host, 2 would be 100001 and so on incrementally." So being that bitnami container has an account user id of `1001` counting forward toward the 1001 places would give the subuid of `101000`. 

I chmod the permissions on the mariadb 'data' directory so that the owner is `101000` like so

`root@redmine:/media/nfs/mariadb_data/mariadb/# chown -R 101000:redmineuser data/`

Then used podman to launch the container and received successful MariaDB output below with no permission errors:

```
mariadb 12:34:16.26 INFO  ==> ** Starting MariaDB **
2022-02-04 12:34:16 0 [Note] /opt/bitnami/mariadb/sbin/mysqld (server 10.6.5-MariaDB) starting as process 1 ...
2022-02-04 12:34:16 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2022-02-04 12:34:16 0 [Note] InnoDB: Number of pools: 1
2022-02-04 12:34:16 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
2022-02-04 12:34:16 0 [Note] mysqld: O_TMPFILE is not supported on /opt/bitnami/mariadb/tmp (disabling future attempts)
2022-02-04 12:34:16 0 [Note] InnoDB: Using Linux native AIO
2022-02-04 12:34:16 0 [Note] InnoDB: Initializing buffer pool, total size = 134217728, chunk size = 134217728
2022-02-04 12:34:16 0 [Note] InnoDB: Completed initialization of buffer pool
2022-02-04 12:34:16 0 [Note] InnoDB: 128 rollback segments are active.
2022-02-04 12:34:16 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2022-02-04 12:34:16 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2022-02-04 12:34:16 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
2022-02-04 12:34:16 0 [Note] InnoDB: 10.6.5 started; log sequence number 30903161; transaction id 39340
2022-02-04 12:34:16 0 [Note] InnoDB: Loading buffer pool(s) from /bitnami/mariadb/data/ib_buffer_pool
2022-02-04 12:34:16 0 [Note] Plugin 'FEEDBACK' is disabled.
2022-02-04 12:34:16 0 [Note] Server socket created on IP: '0.0.0.0'.
2022-02-04 12:34:16 0 [Warning] 'proxies_priv' entry '@% root@redmine' ignored in --skip-name-resolve mode.
2022-02-04 12:34:16 0 [Note] /opt/bitnami/mariadb/sbin/mysqld: ready for connections.
Version: '10.6.5-MariaDB'  socket: '/opt/bitnami/mariadb/tmp/mysql.sock'  port: 3306  Source distribution
2022-02-04 12:34:16 0 [Note] InnoDB: Buffer pool(s) load completed at 220204 12:34:16
```

Long story short for anyone else: Get better at math (for calculating subuid) and check the real subuid number of the podman instances running using `ps` and set the data directory owner to be that number.
